File: ios/Tehilim/Core/Services/PsalmRepository.swift
import Foundation

final class PsalmRepository {
    private let psalmsById: [Int: Psalm]
    let allPsalms: [Psalm]

    init(psalms: [Psalm]) {
        self.allPsalms = psalms.sorted { $0.id < $1.id }
        self.psalmsById = Dictionary(uniqueKeysWithValues: psalms.map { ($0.id, $0) })
    }

    func psalm(id: Int) -> Psalm? { psalmsById[id] }

    func psalms(inBook book: Int) -> [Psalm] {
        guard let range = Psalm.bookRanges[book] else { return [] }
        return allPsalms.filter { range.contains($0.id) }
    }

    func neighbors(of id: Int) -> (prev: Int?, next: Int?) {
        let prev = (1...150).contains(id - 1) ? id - 1 : nil
        let next = (1...150).contains(id + 1) ? id + 1 : nil
        return (prev, next)
    }
}

File: android/app/src/main/java/com/david/tehilim/core/service/PsalmRepository.kt
package com.david.tehilim.core.service

import com.david.tehilim.core.model.Psalm

class PsalmRepository(val allPsalms: List<Psalm>) {

    /** Cherche un Tehilim par son ID (1..150). */
    fun psalm(id: Int): Psalm? = allPsalms.firstOrNull { it.id == id }

    /** Liste des Tehilim du livre donné (1..5). */
    fun psalmsInBook(book: Int): List<Psalm> {
        val range = Psalm.bookRanges[book] ?: return emptyList()
        return allPsalms.filter { range.contains(it.id) }
    }

    /** Voisins (précédent / suivant) dans le corpus complet 1..150. */
    fun neighbors(id: Int): Pair<Int?, Int?> {
        val prev = if (id > 1) id - 1 else null
        val next = if (id < 150) id + 1 else null
        return prev to next
    }
}
